### PR TITLE
feat(builder): Adding support to lock BUILDPACK_URL to a git revision 

### PR DESCRIPTION
**Pull request by amrnt**
You can set an exact version of a buildpack by using a git revision in your BUILDPACK_URL.
`git://repo.git#master` `git://repo.git#v1.2.0`

### CODE REVIEW

**Comment by deis-admin** [comment]
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

File: builder/image/slugbuilder/builder/build.sh [hunk 1]
```diff
@@ -76,7 +76,15 @@ if [[ -n "$BUILDPACK_URL" ]]; then
...
     rm -fr "$buildpack"
-    git clone --quiet --depth=1 "$BUILDPACK_URL" "$buildpack"
+
+    url=${BUILDPACK_URL%#*}
+    branch=${BUILDPACK_URL#*#}
+    if [[ "$branch" == "$url" ]] || [[ -z "${branch}" ]]; then
+        branch="master"
+    fi
+
+    git clone --quiet --branch "$branch" --depth=1 "$url" "$buildpack"
```
Comment (johanneswuerbach) [inline builder/image/slugbuilder/builder/build.sh]: I think you could re-use the `download_buildpack` https://github.com/deis/deis/blob/master/builder/image/slugbuilder/builder/install-buildpacks#L6 function to make this working for branches, tags and specific commits.

Comment (bacongobbler) [inline builder/image/slugbuilder/builder/build.sh]: That function would have to be refactored out into a separate script so it can be used as a shared library

**Comment by bacongobbler** [comment]
I agree with @johanneswuerbach that we could clean 'er up, but this looks ~~great~~awesome!

Code LTGM, though we'll still have to test this manually/test for edge cases :)

File: builder/image/slugbuilder/builder/build.sh (hunk 1 above)
Comment (amrnt) [inline builder/image/slugbuilder/builder/build.sh]: @johanneswuerbach You are absolutely right! I tried to reuse `download_buildpack` but found that it needs more work as @bacongobbler said!

**Comment by carmstrong** [comment]
Code LGTM. This is great - thanks @amrnt!

**Comment by amrnt** [comment]
Thank you guys @carmstrong @bacongobbler :dancer:

**Comment by carmstrong** [comment]
Jenkins, add to whitelist.

**Comment by carmstrong** [comment]
@amrnt I tried this using your example:
```console
$ deis config:set BUILDPACK_URL=git://git@github.com:heroku/heroku-buildpack-ruby.git#v133
Creating config... done, v3

=== pepper-quadrant
DEIS_APP: pepper-quadrant
BUILDPACK_URL: git://git@github.com:heroku/heroku-buildpack-ruby.git#v133
```
```console
$ git push deis master
Counting objects: 115, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (64/64), done.
Writing objects: 100% (115/115), 23.14 KiB | 0 bytes/s, done.
Total 115 (delta 50), reused 99 (delta 43)
-----> Fetching custom buildpack
fatal: I don't handle protocol '"git'
```

**Comment by bacongobbler** [comment]
Something tells me that `git clone` doesn't like the quote escaping.

File: docs/using_deis/using-buildpacks.rst [hunk 2]
```diff
@@ -135,6 +135,14 @@ To use a custom buildpack, set the ``BUILDPACK_URL`` environment variable.
     === humble-autoharp
     BUILDPACK_URL: https://github.com/dpiddy/heroku-buildpack-ruby-minimal
 
+.. note::
+
+    If, however, you're unable to deploy using the latest version of the buildpack,
+    You can set
+    an exact version of a buildpack by using a git revision in your
+    BUILDPACK_URL.
+    git://repo.git#master git://repo.git#v1.2.0
```
Comment (carmstrong) [inline docs/using_deis/using-buildpacks.rst]: Could we change this to something like "For example: " and then render both BUILDPACK_URL and the URL itself as code blocks by using two backticks?

**Comment by amrnt** [comment]
@carmstrong I'm removing all the quote from `git clone` line ( @bacongobbler :+1: ) .

File: docs/using_deis/using-buildpacks.rst (hunk 2 above)
Comment (amrnt) [inline docs/using_deis/using-buildpacks.rst]: @carmstrong To look something like this?

```rst
.. note::

    If, however, you're unable to deploy using the latest version of the buildpack,
    You can set
    an exact version of a buildpack by using a git revision in your
    `BUILDPACK_URL`.
    For example: `BUILDPACK_URL=https://github.com/dpiddy/heroku-buildpack-ruby-minimal#v13`
```

Comment (carmstrong) [inline docs/using_deis/using-buildpacks.rst]: Perfect, except for ReStructured Text formatting, you'll need to use two backticks instead of one whenever you format code.

**Comment by carmstrong** [comment]
@amrnt This still isn't working. Here's the output with debugging enabled:
```console
+ '[' -d /tmp/app ']'
+ cp -rf /tmp/app/. /app
+ cp -r /app/. /tmp/build
+ export APP_DIR=/app
+ APP_DIR=/app
+ export HOME=/app
+ HOME=/app
++ openssl rand -base64 32
+ export REQUEST_ID=YMkwkXybn2luu7FpdnFkrArNNU50IYLqJojjWevhnMI=
+ REQUEST_ID=YMkwkXybn2luu7FpdnFkrArNNU50IYLqJojjWevhnMI=
+ export STACK=cedar-14
+ STACK=cedar-14
+ buildpacks=($buildpack_root/*)
+ selected_buildpack=
+ [[ -n "https://github.com/heroku/heroku-buildpack-ruby.git#v133" ]]
+ echo_title 'Fetching custom buildpack'
+ output_redirect
+ [[ /tmp/slug.tgz == \- ]]
+ cat -
----->' Fetching custom buildpack
-----> Fetching custom buildpack
+ buildpack=/tmp/buildpacks/custom
+ rm -fr /tmp/buildpacks/custom
+ url='"https://github.com/heroku/heroku-buildpack-ruby.git'
+ branch='v133"'
+ [[ v133" == \"\h\t\t\p\s\:\/\/\g\i\t\h\u\b\.\c\o\m\/\h\e\r\o\k\u\/\h\e\r\o\k\u\-\b\u\i\l\d\p\a\c\k\-\r\u\b\y\.\g\i\t ]]
+ [[ -z v133" ]]
+ git clone --quiet --branch 'v133"' --depth=1 '"https://github.com/heroku/heroku-buildpack-ruby.git' /tmp/buildpacks/custom
fatal: I don't handle protocol '"https'
To ssh://git@deis.local3.deisapp.com:2222/oblong-yearling.git
 * [new branch]      master -> master
```

Notice how both `url` and `branch` have an extra quote when the string is interpolated. I'm going to fix this up since I'm in here anyway.

**Comment by carmstrong** [comment]
@amrnt This isn't you - see gh#3078. We'll need to fix that first.

**Comment by carmstrong** [comment]
@amrnt Cam you rebase this on master? We just merged the root cause of the issue. We should also go back to your original code change with quotes around the variables - I think that's the safe way to use them. Thanks!

**Comment by amrnt** [comment]
@carmstrong done!

**Comment by bacongobbler** [comment]
nope :(

```
><> git ci --allow-empty -m 'hi'
[master ead2ce1] hi
><> git push deis master
Warning: Permanently added '[deis.23.253.208.219.xip.io]:2222,[23.253.208.219]:2222' (ECDSA) to the list of known hosts.
Counting objects: 1, done.
Writing objects: 100% (1/1), 183 bytes | 0 bytes/s, done.
Total 1 (delta 0), reused 0 (delta 0)
+ ARGS=3
+ '[' 3 -ne 3 ']'
+ USER=bacongobbler
+ REPO=go.git
+ GIT_SHA=ead2ce1a427d9b73b0aa6a72ba718866cadbb2f8
+ SHORT_SHA=ead2ce1a
+ APP_NAME=go
++ dirname /home/git/builder
+ cd /home/git
++ pwd
+ ROOT_DIR=/home/git
+ DOCKERFILE_SHIM=/home/git/shim.dockerfile
+ REPO_DIR=/home/git/go.git
+ BUILD_DIR=/home/git/go.git/build
+ CACHE_DIR=/home/git/go.git/cache
+ IMAGE_NAME=go:git-ead2ce1a
+ TMP_IMAGE=10.21.12.3:5000/go:git-ead2ce1a
+ mkdir -p /home/git/go.git/build /home/git/go.git/cache
++ mktemp -d --tmpdir=/home/git/go.git/build
+ TMP_DIR=/home/git/go.git/build/tmp.D5gmEru7UN
+ cd /home/git/go.git
+ tar -xmC /home/git/go.git/build/tmp.D5gmEru7UN
+ git archive ead2ce1a427d9b73b0aa6a72ba718866cadbb2f8
+ cd /home/git/go.git/build/tmp.D5gmEru7UN
+ USING_DOCKERFILE=false
+ '[' -f Dockerfile ']'
+ URL=http://10.21.12.1:8000/v1/hooks/config
++ /app/bin/get-app-config -url=http://10.21.12.1:8000/v1/hooks/config -key=t3tYGlmHHGoZWfpzVDs3mXur3L70CsttzCz8CNBIsTzoxLunBgDBbQaZXiMhudRLBg4E3EJcaX3aJ/oRbawb5Q== -user=bacongobbler -app=go
+ RESPONSE='{"owner":"bacongobbler","app":"go","values":{"BUILDPACK_URL":"https://github.com/heroku/heroku-buildpack-go#godep","POWERED_BY":"bacongobbler"},"memory":{},"cpu":{},"tags":{},"uuid":"d7835660-c913-4937-8393-a63e837dd319","created":"2015-02-27T17:56:13UTC","updated":"2015-02-27T17:56:13UTC"}'
+ CODE=0
+ '[' 0 -ne 0 ']'
+ BUILD_OPTS=()
+ BUILD_OPTS+=/usr/bin/docker
+ BUILD_OPTS+=' run -v /etc/environment_proxy:/etc/environment_proxy'
++ echo '{"owner":"bacongobbler","app":"go","values":{"BUILDPACK_URL":"https://github.com/heroku/heroku-buildpack-go#godep","POWERED_BY":"bacongobbler"},"memory":{},"cpu":{},"tags":{},"uuid":"d7835660-c913-4937-8393-a63e837dd319","created":"2015-02-27T17:56:13UTC","updated":"2015-02-27T17:56:13UTC"}'
++ /app/bin/get-app-values
+ BUILD_OPTS+=' -e BUILDPACK_URL='\''https://github.com/heroku/heroku-buildpack-go#godep'\'' -e POWERED_BY='\''bacongobbler'\''
+ '[' '!' -f Dockerfile ']'
+ BUILD_OPTS+=' -d'
+ BUILD_OPTS+=' -v '
++ echo /home/git/go.git/build/tmp.D5gmEru7UN
+ BUILD_OPTS+=/home/git/go.git/build/tmp.D5gmEru7UN
+ BUILD_OPTS+=:/tmp/app
+ BUILD_OPTS+=' -v '
++ echo /home/git/go.git/cache
+ BUILD_OPTS+=/home/git/go.git/cache
+ BUILD_OPTS+=:/tmp/cache:rw
+ chown -R :2000 /home/git/go.git/build/tmp.D5gmEru7UN
+ chown :2000 /home/git/go.git/cache
+ chmod g+rwx /home/git/go.git/build/tmp.D5gmEru7UN
+ BUILD_OPTS+=' deis/slugbuilder'
++ /usr/bin/docker run -v /etc/environment_proxy:/etc/environment_proxy -e 'BUILDPACK_URL='\''https://github.com/heroku/heroku-buildpack-go#godep'\''' -e 'POWERED_BY='\''bacongobbler'\''' -d -v /home/git/go.git/build/tmp.D5gmEru7UN:/tmp/app -v /home/git/go.git/cache:/tmp/cache:rw deis/slugbuilder
+ JOB=7c8a5cade2a8b268e56ebf9e6bf6e9c88e9a45bb84049b067f63e03f338cc159
+ docker attach 7c8a5cade2a8b268e56ebf9e6bf6e9c88e9a45bb84049b067f63e03f338cc159
+ mkdir -p /tmp/cache
+ mkdir -p /tmp/buildpacks
+ mkdir -p /tmp/build/.profile.d
+ '[' -d /tmp/app ']'
+ cp -rf /tmp/app/. /app
+ cp -r /app/. /tmp/build
+ export APP_DIR=/app
+ APP_DIR=/app
+ export HOME=/app+ HOME=/app
++ openssl rand -base64 32
+ export REQUEST_ID=pCoyzRWN0EVA7p/3ssGdTjyWdA9F6sv8Habww0tAH+s=
+ REQUEST_ID=pCoyzRWN0EVA7p/3ssGdTjyWdA9F6sv8Habww0tAH+s=
+ export STACK=cedar-14
+ STACK=cedar-14
+ buildpacks=($buildpack_root/*)
+ selected_buildpack=
+ [[ -n 'https://github.com/heroku/heroku-buildpack-go#godep' ]]
+ echo_title 'Fetching custom buildpack'
----->' Fetching custom buildpack
+ output_redirect
+ [[ /tmp/slug.tgz == \- ]]
+ cat -
-----> Fetching custom buildpack
+ buildpack=/tmp/buildpacks/custom
+ rm -fr /tmp/buildpacks/custom
+ url=''\''https://github.com/heroku/heroku-buildpack-go'
+ branch='godep'\'''
+ '[' 'godep'\''' == ''\''https://github.com/heroku/heroku-buildpack-go' ']'
+ git clone --quiet --branch 'godep'\''' --depth=1 ''\''https://github.com/heroku/heroku-buildpack-go' /tmp/buildpacks/custom
fatal: I don't handle protocol ''https'
To ssh://git@deis.23.253.208.219.xip.io:2222/go.git
   c2a1867..ead2ce1  master -> master
```

**Comment by bacongobbler** [comment]
notice that the following works: `+ [[ -n 'https://github.com/heroku/heroku-buildpack-go#godep' ]]`

So it seems to be an issue with the bash substitution.

**Comment by bacongobbler** [comment]
The following patch works for me:

```
url="$(echo $BUILDPACK_URL | tr -d "'" | sed "s/#.*//")"
branch="$(echo $BUILDPACK_URL | tr -d "'" | sed "s/.*#//")"
```

The `tr -d "'"` is necessary because when you echo $BUILDPACK_URL, you get the literal string `'https://blah.com/asdf/asdf#qwer'` (notice the single quotes). :(

**Comment by bacongobbler** [comment]
note that this is also still an issue in master, so I'll go ahead and patch this up and propose a new PR with your changes. Thanks!

**Comment by amrnt** [comment]
And now I'm quite lost! I thought this issue would be solved by #3129

@carmstrong @bacongobbler Please feel free to close this PR for #3159

**Comment by bacongobbler** [comment]
thanks @amrnt. your commit's in #3159 so your name will still show up in the release notes. Thanks again for taking this up!

**Comment by amrnt** [comment]
:+1: :clap:
